it/src/test/java/com/google/cloud/teleport/it/spanner/SpannerResourceManagerUtilsTest.java

```diff
@@ -0,0 +1,166 @@
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesHyphenWithUnderscore() {
+    String testBaseString = "test-database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesNonLetterFirstCharWithLetter() {
+    String testBaseString = "0_test_database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).matches("[a-z]_test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test_Database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingHyphen() {
+    String testBaseString = "test_database---";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingUnderscore() {
+    String testBaseString = "test_database___";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesDollarSignWithHyphen() {
+    String testBaseString = "test$instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesDotWithHyphen() {
+    String testBaseString = "test.instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesNonLetterFirstCharWithLetter() {
+    String testBaseString = "0-test-instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("[a-z]-test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesUnderscoreWithHyphen() {
+    String testBaseString = "test_instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test-Instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
+  }
+
+  @Test
+  public void testGenerateNewId_inputLengthLongerThanTargetLength() {
+    String longId = "long-test-id-string";
+
+    String actual = generateNewId(longId, 13);
+
+    assertThat(actual.length()).isAtMost(13);
+    assertThat(actual).matches("(\\w|-)*");
```

Comment:
Could we make this regex clearer? Don't we know what the start of the string is as well as what pattern the hash should match?